importer: introduce DDL via internal executor for IMPORT PGDUMP

ZhouXing19 opened the pull request:
This PR continues the work in https://github.com/cockroachdb/cockroach/pull/72317.

---- 

Note: the biggest differences between this PR and #72317 that we change the logic to process DDLs:  _run buffered DDLs (including GRANTS) -> move newly created objs (including the temporary DB) to `OFFLINE` state_ in a single txn.

----

This change is the first of many in the endeavour to run all DDL statement during
IMPORT PGDUMP in a single txn via an internal executor. This is not hooked up to 
IMPORT PGDUMP and will be introduced as a standalone codepath, only to be 
swapped once we have parity with the existing implementation.

During an IMPORT PGDUMP, in the future, our schema parsing phase will:

- Create a temp database in which we will create the schema objects and import
data into them. This database will be invisible to every user except node
via priviliges.

- Parse the dump file, pointing all DDL statements to the new
import database.

- Run all the buffered DDL statements, revoke superusers' privileges on all newly created tables and schemas, run all buffered GRANT statements, and move all newly created tables/schemas to `OFFLINE` state in a single txn using an internal executor.

Note, this change does not add or remove DDL support from PGDUMP but aims
to achieve parity with the existing implementation in `read_import_pgdump.go`.

Follow up work will include:

- [ ] Address remaining TODOs for specific DDLs.
- [ ] Hooking the schema parsing into the data ingestion.
- [ ] Remapping schema objects to the target database from the importing database before we flip the objects to PUBLIC.
- [ ] Cleanup on import job failure.
- [ ] Resumption semantics during the schema parsing phase.
- [ ] Support for ignore_unsupported_statements.

Informs: https://github.com/cockroachdb/cockroach/issues/67076

Link Epic [CRDB-8353](https://cockroachlabs.atlassian.net/browse/CRDB-8353)

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/90084)
<!-- Reviewable:end -->

miretskiy commented on pkg/jobs/jobspb/jobs.proto:
```diff
@@ -479,7 +499,9 @@ message ImportDetails {
   ];
 
-  // next val: 28
+  ProcessDDLStmtStage pgdump_dDL_progress = 28;
```
maybe _ddl_ for proto field name?

knz reviewed (commented):
Don't forget to ask for review / mentorship from someone in the SQL schema team as well.

> Reviewed 1 of 14 files at r1, 1 of 18 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/90084#-:-NFOLkSv4qILABcqhGiS:bcm2i2l)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @rhu713, and @ZhouXing19)

___
*[`pkg/jobs/jobspb/jobs.proto` line 502 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/90084#-NFOLIja9U6wpSm4Oj5j:-NFOLIja9U6wpSm4Oj5k:b-fr230y) ([raw file](https://github.com/cockroachdb/cockroach/blob/4470cadfc91fe961dff85eaf4531fd051fcb0996/pkg/jobs/jobspb/jobs.proto#L502)):*
> ```protobuf
>   ];
> 
>   ProcessDDLStmtStage pgdump_dDL_progress = 28;
> ```

don't use capitals in identifiers. If you want to influence the go identifier name, use a `gogoproto.customname` annotation.


<!-- Sent from Reviewable.io -->

ZhouXing19 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/90084)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @knz, @miretskiy, and @rhu713)

___
*[`pkg/jobs/jobspb/jobs.proto` line 502 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/90084#-NEqnWRwgwpJ-QRokL1L-r2-502:-NFPBkQUDM-Fuc67bVMF:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ade837d50d65d4b74e37143efc47a1cf322182f/pkg/jobs/jobspb/jobs.proto#L502)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

maybe _ddl_ for proto field name?
</blockquote></details>

Done.
___
*[`pkg/jobs/jobspb/jobs.proto` line 502 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/90084#-NFOLIja9U6wpSm4Oj5j:-NFPBku23StAfJsE9sq5:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4470cadfc91fe961dff85eaf4531fd051fcb0996/pkg/jobs/jobspb/jobs.proto#L502)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

don't use capitals in identifiers. If you want to influence the go identifier name, use a `gogoproto.customname` annotation.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

ZhouXing19 commented:
Coming back to the TODO about privilege on database:

>  // TODO(adityamaru): Figure out how to create the database descriptor with
// privileges for only the node user. Currently, root and admin have ALL
// privileges on the database descriptor. This is enforced by
// ValidateSuperuserPrivileges when writing the descriptor to store.
// I tried moving the database descriptor to OFFLINE instead of mucking with
// privileges, but this prevents us from running any DDL statements on this
// database.

Per an offline convo with @adityamaru, we think the privileges on the newly created db/schemas/tables doesn't really matter, since we can now {run all DDLs and GRANTS, move all objects to `OFFLINE`} in a single txn. The grantee won't be able to insert into the table / create new table anyways. Thus, it should be sufficient to move all newly created db/schemas/tables to `OFFLINE`. 
Corresponding tests are added in `TestPrivilegeOnTempDB`.

rafiss reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/90084)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @knz, @miretskiy, @rhu713, and @ZhouXing19)

___
*[`pkg/jobs/jobspb/jobs.proto` line 408 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/90084#-NFosVK70m34uIpucTmt:-NFosVK70m34uIpucTmu:b-o6a21q) ([raw file](https://github.com/cockroachdb/cockroach/blob/92fa502c0d65c967ce36b36e9c36d51a32a0a990/pkg/jobs/jobspb/jobs.proto#L408)):*
> ```protobuf
> }
> 
> // The ProcessDDLStmtStage is consist of 3 major steps:
> ```

nit: "is consist of" -> "consists of"
___
*[`pkg/sql/importer/import_pgdump.go` line 42 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/90084#-NFosABY9ZhRhib4JZP7:-NFosABY9ZhRhib4JZP8:bkd58gu) ([raw file](https://github.com/cockroachdb/cockroach/blob/92fa502c0d65c967ce36b36e9c36d51a32a0a990/pkg/sql/importer/import_pgdump.go#L42)):*
> ```Go
> 	// ImportTempPgdumpDB is the default name of the temp database for newly
> 	// created schema from DDL stmts in the dump file.
> 	ImportTempPgdumpDB = "crdb_temp_pgdump_import"
> ```

nit: fix the uppercase in this variable. perhaps like `TempImportPGDumpDBName`
___
*[`pkg/sql/importer/import_pgdump.go` line 67 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/90084#-NFosR6U3Pv_SGhpQYyS:-NFosR6U3Pv_SGhpQYyT:bque7z2) ([raw file](https://github.com/cockroachdb/cockroach/blob/92fa502c0d65c967ce36b36e9c36d51a32a0a990/pkg/sql/importer/import_pgdump.go#L67)):*
> ```Go
> }
> 
> // PostgresDDLHandler stored DDL stmts and the temp db to store the new schemas.
> ```

nit: stored -> stores
___
*[`pkg/sql/importer/import_pgdump.go` line 74 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/90084#-NFousIF4OLQdLKBbNLe:-NFousIF4OLQdLKBbNLf:b-eqn2ni) ([raw file](https://github.com/cockroachdb/cockroach/blob/92fa502c0d65c967ce36b36e9c36d51a32a0a990/pkg/sql/importer/import_pgdump.go#L74)):*
> ```Go
> }
> 
> func formatPostgresStatement(n tree.NodeFormatter) string {
> ```

nit: don't think we need `Postgres` in the name. also, the param can be a `tree.Statement`
___
*[`pkg/sql/importer/import_pgdump.go` line 83 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/90084#-NFov3La9s9CbdKjIEJu:-NFov3La9s9CbdKjIEJv:bslu5se) ([raw file](https://github.com/cockroachdb/cockroach/blob/92fa502c0d65c967ce36b36e9c36d51a32a0a990/pkg/sql/importer/import_pgdump.go#L83)):*
> ```Go
> }
> 
> func rewritePostgresStatementTableName(
> ```

nit: don't need postgres in the name
___
*[`pkg/sql/importer/import_pgdump.go` line 84 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/90084#-NFouli68yPiXDPU0jIv:-NFouli68yPiXDPU0jIw:b-rfkfsk) ([raw file](https://github.com/cockroachdb/cockroach/blob/92fa502c0d65c967ce36b36e9c36d51a32a0a990/pkg/sql/importer/import_pgdump.go#L84)):*
> ```Go
> 
> func rewritePostgresStatementTableName(
> 	n tree.NodeFormatter, dumpDatabaseName string,
> ```

nit: the parameter here should be a `tree.Statement` not a `NodeFormatter`
___
*[`pkg/sql/importer/import_pgdump.go` line 87 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/90084#-NFowt4h0tgKgDGp2qxg:-NFowt4h0tgKgDGp2qxh:bh5wl12) ([raw file](https://github.com/cockroachdb/cockroach/blob/92fa502c0d65c967ce36b36e9c36d51a32a0a990/pkg/sql/importer/import_pgdump.go#L87)):*
> ```Go
> ) (string, error) {
> 	var err error
> 	f := tree.NewFmtCtx(
> ```

i don't think FmtCtx is a good fit for what this function is trying to do. it's a bit confusing, since it also duplicates some of the responsibility of the function right above this one.

instead, i think this function should _only_ be in in charge of changing the catalog name for every name that appears in the statement, and it should not do any formatting at all.

the impl could be something like:
```
replaceFn := func(expr Expr) (recurse bool, newExpr Expr, err error) {
  if _, ok := expr.(tree.TableName); ok {
    // replace Catalog if needed
    return false, newTableName, nil
  }
  // might also need logic for other types of names like ObjectName
}
tree.SimpleStmtVisit(stmt, replaceFn)
```

and make sure all the statements you care about implement the `walkableStmt` interface: https://github.com/cockroachdb/cockroach/blob/02ff871a4b4e16f7082298d7e098bfed70f528a9/pkg/sql/sem/tree/walk.go#L1663
___
*[`pkg/sql/importer/import_pgdump.go` line 102 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/90084#-NFotCi0AGJbedzfVaYv:-NFotCi14Yh9c2BVmGIM:b-ozeija) ([raw file](https://github.com/cockroachdb/cockroach/blob/92fa502c0d65c967ce36b36e9c36d51a32a0a990/pkg/sql/importer/import_pgdump.go#L102)):*
> ```Go
> 			}
> 
> 			fmtctx.WithReformatTableNames(nil, func() {
> ```

something is off here - you don't need to call both `FmtReformatTableNames` and also `WithReformatTableNames`. in this case, you should just remove the call to `WithReformatTableNames` entirely.
___
*[`pkg/sql/importer/import_pgdump.go` line 117 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/90084#-NFouPwg9Qsb_mv1QyLn:-NFouPwiB7euZbFWYxXM:b-4idmtz) ([raw file](https://github.com/cockroachdb/cockroach/blob/92fa502c0d65c967ce36b36e9c36d51a32a0a990/pkg/sql/importer/import_pgdump.go#L117)):*
> ```Go
> 	ctx context.Context,
> 	evalCtx *eval.Context,
> 	postgresStmt interface{},
> ```

can this be `tree.Statement` instead of `interface{}`
___
*[`pkg/sql/importer/import_pgdump.go` line 123 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/90084#-NFoycVr3FxxVilzkGn7:-NFoycVr3FxxVilzkGn8:b-5ixrdu) ([raw file](https://github.com/cockroachdb/cockroach/blob/92fa502c0d65c967ce36b36e9c36d51a32a0a990/pkg/sql/importer/import_pgdump.go#L123)):*
> ```Go
> ) error {
> 	switch stmt := postgresStmt.(type) {
> 	case *tree.CreateRole:
> ```

needing special logic for each statement isn't ideal. instead, we can use `tree.SimpleStmtVisit` as mentioned above, and _always_ call the replaceName function on every statement
___
*[`pkg/sql/importer/import_pgdump.go` line 255 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/90084#-NFozLYQDiv4nRVcirUc:-NFozLYQDiv4nRVcirUd:b-2eiqci) ([raw file](https://github.com/cockroachdb/cockroach/blob/92fa502c0d65c967ce36b36e9c36d51a32a0a990/pkg/sql/importer/import_pgdump.go#L255)):*
> ```Go
> 					}
> 					// Now that we have all of the datums, we can execute the overload.
> 					fnSQL, err := fn(ctx, evalCtx, datums)
> ```

i don't quite follow this function eval part. why are we evaluating the function? also, it seems like something that could be pulled into its own function so it can have better comments
___
*[`pkg/sql/importer/import_pgdump.go` line 296 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/90084#-NFozbt8BobiyXnQ0cNQ:-NFozbt8BobiyXnQ0cNR:b8xvhdb) ([raw file](https://github.com/cockroachdb/cockroach/blob/92fa502c0d65c967ce36b36e9c36d51a32a0a990/pkg/sql/importer/import_pgdump.go#L296)):*
> ```Go
> 			if err := p.ExecCfg().DB.Txn(ctx, func(ctx context.Context, txn *kv.Txn) error {
> 				descsCol := p.ExtendedEvalContext().Descs
> 				err := descsCol.Direct().CheckObjectCollision(
> ```

is this the right way to check for a table with a name? i think there are other catalog or descs interfaces that are preferred, but check with sql-schema
___
*[`pkg/sql/importer/import_pgdump.go` line 343 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/90084#-NFp2n8F53U7WN9H4eGP:-NFp2n8F53U7WN9H4eGQ:b-b5qo0f) ([raw file](https://github.com/cockroachdb/cockroach/blob/92fa502c0d65c967ce36b36e9c36d51a32a0a990/pkg/sql/importer/import_pgdump.go#L343)):*
> ```Go
> 		}
> 		for _, table := range stmt.Targets.Tables.TablePatterns {
> 			explicitCatalog, _ := table.GetExplicitCatalog()
> ```

why is the second return val ignored? it seems like we don't need it at all then
___
*[`pkg/sql/importer/import_pgdump.go` line 441 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/90084#-NFp2zOF67xgtRCLYGQi:-NFp2zOF67xgtRCLYGQj:b-450z1k) ([raw file](https://github.com/cockroachdb/cockroach/blob/92fa502c0d65c967ce36b36e9c36d51a32a0a990/pkg/sql/importer/import_pgdump.go#L441)):*
> ```Go
> // MoveObjectsInTempDatabaseToState is to move the db and the objects in it to
> // a certain state.
> func MoveObjectsInTempDatabaseToState(
> ```

i'm not so sure about the logic for taking tables offline and modifying the descriptors, so this part should be reviewed by sql-schema
___
*[`pkg/sql/sem/tree/format.go` line 308 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/90084#-NFp0G58EqrQ7nv6GA1V:-NFp0G58EqrQ7nv6GA1W:ba25sm0) ([raw file](https://github.com/cockroachdb/cockroach/blob/92fa502c0d65c967ce36b36e9c36d51a32a0a990/pkg/sql/sem/tree/format.go#L308)):*
> ```Go
> 
> // FmtReformatTableNames modifies FmtCtx to to substitute the printing of
> // tableNameFmtParsable using the provided function.
> ```

this should say "substitute the printing of table names using the provided function"
___
*[`pkg/sql/sem/tree/table_pattern.go` line 44 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/90084#-NFp2Vwg80NHEFwcXmXJ:-NFp2Vwl0GdOGEkoFSd9:bv6klxz) ([raw file](https://github.com/cockroachdb/cockroach/blob/92fa502c0d65c967ce36b36e9c36d51a32a0a990/pkg/sql/sem/tree/table_pattern.go#L44)):*
> ```Go
> 	// GetExplicitCatalog returns the catalog specified on the object, if it exists.
> 	// If not, returns a not-found error.
> 	GetExplicitCatalog() (Name, error)
> ```

i think it would be better to return a `bool` here instead of an `error`
___
*[`pkg/sql/sem/tree/table_pattern.go` line 47 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/90084#-NFp2e3QD5YB5mcff-8_:-NFp2e3QD5YB5mcff-8a:b-c5jhea) ([raw file](https://github.com/cockroachdb/cockroach/blob/92fa502c0d65c967ce36b36e9c36d51a32a0a990/pkg/sql/sem/tree/table_pattern.go#L47)):*
> ```Go
> 
> 	// SetExplicitCatalog sets the explicit catalog if it already exists.
> 	SetExplicitCatalog(Name) error
> ```

i don't see a need to return an error here. it can just return a bool if it set anything, or it could even return nothing.


<!-- Sent from Reviewable.io -->

ZhouXing19 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/90084)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @knz, @miretskiy, @rafiss, and @rhu713)

___
*[`pkg/jobs/jobspb/jobs.proto` line 408 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/90084#-NFosVK70m34uIpucTmt:-NFp8Y0vAs5Wmwgda9ey:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/92fa502c0d65c967ce36b36e9c36d51a32a0a990/pkg/jobs/jobspb/jobs.proto#L408)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: "is consist of" -> "consists of"
</blockquote></details>

Done.
___
*[`pkg/sql/importer/import_pgdump.go` line 42 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/90084#-NFosABY9ZhRhib4JZP7:-NFp8Wtb8ciF2yH6XcwA:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/92fa502c0d65c967ce36b36e9c36d51a32a0a990/pkg/sql/importer/import_pgdump.go#L42)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: fix the uppercase in this variable. perhaps like `TempImportPGDumpDBName`
</blockquote></details>

Done.
___
*[`pkg/sql/importer/import_pgdump.go` line 67 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/90084#-NFosR6U3Pv_SGhpQYyS:-NFp8VPtAzO2XYC9-LYP:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/92fa502c0d65c967ce36b36e9c36d51a32a0a990/pkg/sql/importer/import_pgdump.go#L67)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: stored -> stores
</blockquote></details>

Done.
___
*[`pkg/sql/importer/import_pgdump.go` line 74 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/90084#-NFousIF4OLQdLKBbNLe:-NFp8WBX8CO48zZKMzNW:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/92fa502c0d65c967ce36b36e9c36d51a32a0a990/pkg/sql/importer/import_pgdump.go#L74)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: don't think we need `Postgres` in the name. also, the param can be a `tree.Statement`
</blockquote></details>

Done.
___
*[`pkg/sql/importer/import_pgdump.go` line 83 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/90084#-NFov3La9s9CbdKjIEJu:-NFp8f7J2SoMD5bWAwBY:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/92fa502c0d65c967ce36b36e9c36d51a32a0a990/pkg/sql/importer/import_pgdump.go#L83)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: don't need postgres in the name
</blockquote></details>

Done.



_Code quote:_
```Go
rewritePostgresStatementTableName
```
___
*[`pkg/sql/importer/import_pgdump.go` line 84 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/90084#-NFouli68yPiXDPU0jIv:-NFp8fgkDoCV3W2Uxj6w:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/92fa502c0d65c967ce36b36e9c36d51a32a0a990/pkg/sql/importer/import_pgdump.go#L84)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: the parameter here should be a `tree.Statement` not a `NodeFormatter`
</blockquote></details>

Done.
___
*[`pkg/sql/importer/import_pgdump.go` line 87 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/90084#-NFowt4h0tgKgDGp2qxg:-NFpIZJ_89ovCZ3iAW21:bvejl9f) ([raw file](https://github.com/cockroachdb/cockroach/blob/92fa502c0d65c967ce36b36e9c36d51a32a0a990/pkg/sql/importer/import_pgdump.go#L87)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i don't think FmtCtx is a good fit for what this function is trying to do. it's a bit confusing, since it also duplicates some of the responsibility of the function right above this one.

instead, i think this function should _only_ be in in charge of changing the catalog name for every name that appears in the statement, and it should not do any formatting at all.

the impl could be something like:
```
replaceFn := func(expr Expr) (recurse bool, newExpr Expr, err error) {
  if _, ok := expr.(tree.TableName); ok {
    // replace Catalog if needed
    return false, newTableName, nil
  }
  // might also need logic for other types of names like ObjectName
}
tree.SimpleStmtVisit(stmt, replaceFn)
```

and make sure all the statements you care about implement the `walkableStmt` interface: https://github.com/cockroachdb/cockroach/blob/02ff871a4b4e16f7082298d7e098bfed70f528a9/pkg/sql/sem/tree/walk.go#L1663
</blockquote></details>

Ah TIL, it does look much better.

But it seems to me that `tree.Expr` are implemented by "expression" types, such as `BinaryExpr`, `CollateExpr`, `DDate` and `DArray`. The specific part of statement, such as `tree.TableName` does not implement it. I think it was intentional that we have this distinction, right ... ?

I can do a weird conversion like `tn, ok := expr.(tree.NodeFormatter).(*tree.TableName)` but eventually I had to convert this new `tree.TableName` back to `tree.Expr`, which seems not very correct to me. 

Or rather, should we instead change 

```
type SimpleVisitFn func(expr Expr) (recurse bool, newExpr Expr, err error)
```

to 

```
type SimpleVisitFn func(n NodeFormatter) (recurse bool, n NodeFormatter, err error)
```
?
___
*[`pkg/sql/importer/import_pgdump.go` line 102 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/90084#-NFotCi0AGJbedzfVaYv:-NFpAT9zCs5dqUfS40gK:b-lr3wl) ([raw file](https://github.com/cockroachdb/cockroach/blob/92fa502c0d65c967ce36b36e9c36d51a32a0a990/pkg/sql/importer/import_pgdump.go#L102)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

something is off here - you don't need to call both `FmtReformatTableNames` and also `WithReformatTableNames`. in this case, you should just remove the call to `WithReformatTableNames` entirely.
</blockquote></details>

TIL! Done.
___
*[`pkg/sql/importer/import_pgdump.go` line 117 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/90084#-NFouPwg9Qsb_mv1QyLn:-NFpAYRpAdf7uUwpaOsa:b4gsnsc) ([raw file](https://github.com/cockroachdb/cockroach/blob/92fa502c0d65c967ce36b36e9c36d51a32a0a990/pkg/sql/importer/import_pgdump.go#L117)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

can this be `tree.Statement` instead of `interface{}`
</blockquote></details>

Done.
But for more context, we were using `interface{}` to be compatible with stmts of `importer.copyData` type. Seems to me that it's used to store PGCOPY data for `IMPORT INTO`. But since we no longer support PGCOPY for `IMPORT INTO`, does it mean that we should disallow `copyData` here? 
Maybe @adityamaru can help confirm?  
___
*[`pkg/sql/importer/import_pgdump.go` line 343 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/90084#-NFp2n8F53U7WN9H4eGP:-NFpBo7U3au-zVDeAtUI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/92fa502c0d65c967ce36b36e9c36d51a32a0a990/pkg/sql/importer/import_pgdump.go#L343)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

why is the second return val ignored? it seems like we don't need it at all then
</blockquote></details>

Done.
___
*[`pkg/sql/sem/tree/format.go` line 308 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/90084#-NFp0G58EqrQ7nv6GA1V:-NFpJsUK0khaZ70ESoTo:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/92fa502c0d65c967ce36b36e9c36d51a32a0a990/pkg/sql/sem/tree/format.go#L308)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

this should say "substitute the printing of table names using the provided function"
</blockquote></details>

Done.
___
*[`pkg/sql/sem/tree/table_pattern.go` line 44 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/90084#-NFp2Vwg80NHEFwcXmXJ:-NFpBxaa5c7NXxX4Igye:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/92fa502c0d65c967ce36b36e9c36d51a32a0a990/pkg/sql/sem/tree/table_pattern.go#L44)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i think it would be better to return a `bool` here instead of an `error`
</blockquote></details>

Done.
___
*[`pkg/sql/sem/tree/table_pattern.go` line 47 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/90084#-NFp2e3QD5YB5mcff-8_:-NFpD06O8I5K_GOjofZt:b-4izawp) ([raw file](https://github.com/cockroachdb/cockroach/blob/92fa502c0d65c967ce36b36e9c36d51a32a0a990/pkg/sql/sem/tree/table_pattern.go#L47)):*
Changed it to return a bool -- I think it's easier to debug in the future if this step actually fails (for in `UnresolvedName.SetExplicitCatalog()` we have `u.NormalizeTablePattern()` that may error out.)

> // SetExplicitCatalog sets the explicit catalog if it already exists.
func (u *UnresolvedName) SetExplicitCatalog(n Name) bool {
	pattern, err := u.NormalizeTablePattern()
	if err != nil {
		return false
	}
	return pattern.SetExplicitCatalog(n)
}


<!-- Sent from Reviewable.io -->

ZhouXing19 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/90084)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @knz, @miretskiy, @rafiss, and @rhu713)

___
*[`pkg/sql/importer/import_pgdump.go` line 343 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/90084#-NFp2n8F53U7WN9H4eGP:-NFpTfL5DeN1QhD_09ZN:b-918lmt) ([raw file](https://github.com/cockroachdb/cockroach/blob/92fa502c0d65c967ce36b36e9c36d51a32a0a990/pkg/sql/importer/import_pgdump.go#L343)):*
<details><summary><i>Previously, ZhouXing19 (Jane Xing) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Changed it to check the returned bool.


<!-- Sent from Reviewable.io -->

ZhouXing19 commented:
*[`pkg/sql/importer/import_pgdump.go` line 102 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/90084#-NFotCi0AGJbedzfVaYv:-NFxyDwnBbEh1-LYD7V-:b-p47ma3) ([raw file](https://github.com/cockroachdb/cockroach/blob/92fa502c0d65c967ce36b36e9c36d51a32a0a990/pkg/sql/importer/import_pgdump.go#L102)):*
<details><summary><i>Previously, ZhouXing19 (Jane Xing) wrote…</i></summary><blockquote>

TIL! Done.
</blockquote></details>

Seems that we do need `WithReformatTableNames` here. It's because <> leaves it to `tableNameFormatter` to write the table name to the buffer, if `tableNameFormatter` is not nil.

``` go
// Format implements the NodeFormatter interface.
func (t *TableName) Format(ctx *FmtCtx) {
	if ctx.tableNameFormatter != nil {
		ctx.tableNameFormatter(ctx, t)
		return
	}
	t.ObjectNamePrefix.Format(ctx)
	if t.ExplicitSchema || ctx.alwaysFormatTablePrefix() {
		ctx.WriteByte('.')
	}
	ctx.FormatNode(&t.ObjectName)
}
```
Per other usages of `FmtReformatTableNames()`,  there's always a `WithReformatTableNames` called in the `tableNameFormatter`. 

But indeed this is not very intuitive. Maybe we should fix `TableName.Format` so that all writes eventually go through `ctx.FormatNode(&t.ObjectName)`?


<!-- Sent from Reviewable.io -->

ZhouXing19 commented:
*[`pkg/sql/importer/import_pgdump.go` line 102 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/90084#-NFotCi0AGJbedzfVaYv:-NFxzKQE8rlcii2iV07l:b9rcsvo) ([raw file](https://github.com/cockroachdb/cockroach/blob/92fa502c0d65c967ce36b36e9c36d51a32a0a990/pkg/sql/importer/import_pgdump.go#L102)):*
<details><summary><i>Previously, ZhouXing19 (Jane Xing) wrote…</i></summary><blockquote>

Seems that we do need `WithReformatTableNames` here. It's because <> leaves it to `tableNameFormatter` to write the table name to the buffer, if `tableNameFormatter` is not nil.

``` go
// Format implements the NodeFormatter interface.
func (t *TableName) Format(ctx *FmtCtx) {
	if ctx.tableNameFormatter != nil {
		ctx.tableNameFormatter(ctx, t)
		return
	}
	t.ObjectNamePrefix.Format(ctx)
	if t.ExplicitSchema || ctx.alwaysFormatTablePrefix() {
		ctx.WriteByte('.')
	}
	ctx.FormatNode(&t.ObjectName)
}
```
Per other usages of `FmtReformatTableNames()`,  there's always a `WithReformatTableNames` called in the `tableNameFormatter`. 

But indeed this is not very intuitive. Maybe we should fix `TableName.Format` so that all writes eventually go through `ctx.FormatNode(&t.ObjectName)`?
</blockquote></details>

For the `<>` above I meant `TableName.Format()` :sweat_smile: 


<!-- Sent from Reviewable.io -->

ZhouXing19 committed:
sql/importer: introduce DDL via internal executor for IMPORT PGDUMP

This PR is to continue the work in #72317.

This change is the first of many in the endeavour to run all DDL statement during
IMPORT PGDUMP via an internal executor. This is not hooked up to IMPORT PGDUMP and
will be introduced as a standalone codepath, only to be swapped once we have parity
with the existing implementation.

During an IMPORT PGDUMP, in the future, our schema parsing phase will:

- Create a temp database in which we will create the schema objects and import
data into them. This database will be invisible to every user except node
via priviliges.

- Parse the dump file, pointing all DDL statements to the new
import database.

- Run {all the buffered DDL statements (including GRANTS), and move all newly created
objects to OFFLINE state} in a single txn using an internal executor.

Note, this change does not add or remove DDL support from PGDUMP but aims
to achieve parity with the existing implementation in `read_import_pgdump.go`.

Informs: #67076

Co-authored with Aditya Maru <adityamaru@gmail.com>.

Release note: None

ZhouXing19 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/90084)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @knz, @miretskiy, @rafiss, and @rhu713)

___
*[`pkg/sql/importer/import_pgdump.go` line 343 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/90084#-NFp2n8F53U7WN9H4eGP:-NFyqU1M09cXG1Ede-0a:bfk0h6v) ([raw file](https://github.com/cockroachdb/cockroach/blob/92fa502c0d65c967ce36b36e9c36d51a32a0a990/pkg/sql/importer/import_pgdump.go#L343)):*
<details><summary><i>Previously, ZhouXing19 (Jane Xing) wrote…</i></summary><blockquote>

Changed it to check the returned bool.
</blockquote></details>

I now just convert it to `tree.TableName`. It's simpler and no need to add new method to the `TablePattern` interface.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> This is not hooked up to
IMPORT PGDUMP and will be introduced as a standalone codepath, only to be
swapped once we have parity with the existing implementation.

Have you tried it and do you understand the differences? Is there a way we can hook it up and then allow-list the tests which fail? I think the basic ideas are fine. I need to take a closer look at some of the descriptor management. I worry about many of the arbitrary restrictions we have around things like `ALTER TABLE ... ADD PRIMARY KEY` and other schema changes. I suspect that this approach is going to be great in the long run, and it's going to take a long time to get fully there (probably as long as it takes the schema team to get full coverage for things in the declarative schema changer, which is going to be quite some time). 

If I have any advice, it'd be to actually hook this up in a way where you can play with it and iterate on things that don't work. Getting this close without hooking it up completely and running more involved tests seems like a shame. I think it'd be fine to hook it up to the point where you never run the data ingestion, you just no-op at that point and then move the things to public. That will give you a real feel for how this works from a schema only perspective.

I'll also say that nothing beats data-driven or logictest format for the end-to-end stuff. I notice that the existing testing seems pretty weak around this code. I think one of the best bang-for-your-buck things you could do is set up an ergonomic testing suite where it's easy to add cases and easy to verify parity between your implementation and the old implementation. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/90084#-:-NGPSGth2lrm1QQ9ZhRF:bi7lthb)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @knz, @miretskiy, @rafiss, @rhu713, and @ZhouXing19)

___
*[`pkg/sql/catalog/descs/collection.go` line 534 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/90084#-NGPRQsXAmKyJr-odY7a:-NGPRQsXAmKyJr-odY7b:b7gokad) ([raw file](https://github.com/cockroachdb/cockroach/blob/048f2a1a56177111648cca7eb573469b15b4db36/pkg/sql/catalog/descs/collection.go#L534)):*
> ```Go
> 	ctx context.Context, txn *kv.Txn, db catalog.DatabaseDescriptor,
> ) ([]catalog.SchemaDescriptor, error) {
> 	all, err := tc.GetAllDescriptors(ctx, txn)
> ```

I'd prefer if you used existing logic to get all the schema IDs and then fetched them all by ID in a batch. Uses of `GetAllDescriptors` have a history of causing problems. 
___
*[`pkg/sql/importer/import_pgdump.go` line 441 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/90084#-NFp2zOF67xgtRCLYGQi:-NGPUtI73A-XClSAD1g3:boz362v) ([raw file](https://github.com/cockroachdb/cockroach/blob/92fa502c0d65c967ce36b36e9c36d51a32a0a990/pkg/sql/importer/import_pgdump.go#L441)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i'm not so sure about the logic for taking tables offline and modifying the descriptors, so this part should be reviewed by sql-schema
</blockquote></details>

this itself isn't really off the mark
___
*[`pkg/sql/importer/import_pgdump.go` line 289 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/90084#-NGPRnRA4t0wcDze0Qp8:-NGPRnRA4t0wcDze0Qp9:b-aygejn) ([raw file](https://github.com/cockroachdb/cockroach/blob/048f2a1a56177111648cca7eb573469b15b4db36/pkg/sql/importer/import_pgdump.go#L289)):*
> ```Go
> 		// into and same public schema, then we throw an error telling the user to
> 		// drop the conflicting existing table to proceed.
> 		// Otherwise, we silently ignore the drop statement and continue with the import.
> ```

what if it drops a table created earlier in this pgdump file? 
___
*[`pkg/sql/importer/import_pgdump.go` line 448 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/90084#-NGPUvX45cdqSG5UicZc:-NGPUvX45cdqSG5UicZd:b-jham7p) ([raw file](https://github.com/cockroachdb/cockroach/blob/048f2a1a56177111648cca7eb573469b15b4db36/pkg/sql/importer/import_pgdump.go#L448)):*
> ```Go
> ) (*dbdesc.Mutable, []*tabledesc.Mutable, []*schemadesc.Mutable, error) {
> 	importedTables := make([]*tabledesc.Mutable, 0)
> 	importedSchemas := make([]*schemadesc.Mutable, 0)
> ```

nit: you don't need to provide `0`, but also if you're not initializing a slice, you can still append to it. I think `var importedTables *[]tabledesc.Mutable` or naming the return would be cleaner. 



_Code quote:_
```Go
	importedTables := make([]*tabledesc.Mutable, 0)
	importedSchemas := make([]*schemadesc.Mutable, 0)
```
___
*[`pkg/sql/importer/import_pgdump.go` line 555 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/90084#-NGPS2c7AZ_M_zPKFHOR:-NGPS2c7AZ_M_zPKFHOS:bqy02dp) ([raw file](https://github.com/cockroachdb/cockroach/blob/048f2a1a56177111648cca7eb573469b15b4db36/pkg/sql/importer/import_pgdump.go#L555)):*
> ```Go
> 	) error {
> 		// Run the buffered DDL statements.
> 		if err := runDDLStmtsFromDumpFile(ctx, h.BufferedDDLStmts, txn, ie); err != nil {
> ```

I'm curious to what extent this all works. In theory, it's a very smart idea. My primary concern is that a lot of schema changes create jobs and those jobs I think are going to be rather unhappy about the tables being offline. 


<!-- Sent from Reviewable.io -->

postamar commented on pkg/sql/importer/import_pgdump.go:
```diff
@@ -0,0 +1,581 @@
+	tableDescs []*tabledesc.Mutable,
+	schemaDescs []*schemadesc.Mutable,
+	err error,
+) {
```
Same comment as elsewhere: there exist descriptor kinds other than databases, tables and schemas. That may not be true now but it will be in the future, so you should probably return `([]catalog.Descriptor, error)` instead. 

They certainly shouldn't be exposed as mutable descriptors, use the `.ImmutableCopy()` method.

postamar commented on pkg/sql/importer/import_pgdump.go:
```diff
@@ -0,0 +1,581 @@
+// TODO(janexing): Figure out job resumption semantics, via
+// ImportDetail.PgdumpDDLProgress.
+// TODO(janexing): Update job status to reflect stage.
+func ProcessDDLStatements(
```
Is it possible to use this for real and not just in unit tests? Otherwise, how do we know if it actually works? If it only works some of the time, is it possible to clean up and fall back to the old import logic (assuming there is one) when this throws an error?

postamar commented on pkg/sql/importer/import_pgdump.go:
```diff
@@ -0,0 +1,581 @@
+				switch ast := fnStmt.AST.(type) {
+				case *tree.CreateTable, *tree.CreateIndex, *tree.CreateSequence:
+					return errors.WithHintf(wrappedErr, "is the target of the db.target form? "+
+						"please use db.schema.target or schema.target format for target in %q", ast.String())
```
Is this hint actionable? I don't know much about pgdump files but I imagine that users can't actually edit them.

postamar commented on pkg/sql/importer/import_pgdump.go:
```diff
@@ -0,0 +1,581 @@
+	p sql.JobExecContext,
+	parentID descpb.ID,
+	handler *PostgresDDLHandler,
+) error {
```
Should this function or its parent check for context cancellation? I ask because I imagine that pgdumps might be quite big and take significant time to process.

In a similar vein, should we log the statements? Perhaps only when `log.ExpensiveLogEnabled`?

postamar commented on pkg/sql/importer/import_pgdump.go:
```diff
@@ -0,0 +1,581 @@
+			return stmt
+		}
+	default:
+		return wrapErrorWithUnsupportedHint(errors.Errorf("unsupported %T statement: %s", stmt, stmt))
```
How complete is `bufferDDLPostgresStatement`? Does it cover all of the statements that may be found in a pgdump, and therefore this `default` case is to guard against something completely unexpected? Or is this just a subset?

postamar commented on pkg/sql/importer/import_pgdump.go:
```diff
@@ -0,0 +1,581 @@
+		// If we find a table with the same name in the target DB we are importing
+		// into and same public schema, then we throw an error telling the user to
+		// drop the conflicting existing table to proceed.
+		// Otherwise, we silently ignore the drop statement and continue with the import.
```
This is probably my ignorance showing but I find it odd that pgdump files would ever contain conflicting table names. If this is not expected, why not let the statements play out in the transaction and error there?

postamar commented on pkg/sql/importer/import_pgdump_test.go:
```diff
@@ -0,0 +1,472 @@
+// TestParseDDLStatementsFromDumpFile tests that DDL statements in a dump file
+// are correctly parsed, and the database name for fully-qualified objects is
+// replaced with the temporary database being imported into.
+func TestParseDDLStatementsFromDumpFile(t *testing.T) {
```
Like Andrew suggested I can't encourage you enough to port these unit tests to the datadriven framework. It's really not as hard to do as it seems and it makes it so much easier to add test cases later on.

postamar commented on pkg/jobs/jobspb/jobs.proto:
```diff
@@ -479,7 +499,9 @@ message ImportDetails {
   ];
 
-  // next val: 28
+  ProcessDDLStmtStage pgdump_ddl_progress = 28;
```
nit: add a `(gogoproto.customname)` annotation?

postamar reviewed (commented):
I'm also curious to which extent any of this works. You'll probably be better off finding out earlier rather than later.

Would it be hard to generate a few dozen smaller, fairly realistic pgdumps, check them into this repo, and exercise this code against them? The expected output might be the result of a SHOW CREATE ALL TABLES.

ZhouXing19 commented:
Thanks for reviewing! I'm working on adding the data-driven test framework and trying to hook this up with the existing IMPORT workflow.

ZhouXing19 commented:
I will pause on this PR as the team is focusing on other priorities for the release.
